express-notes-api/index.js

```diff
@@ -0,0 +1,110 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+app.use(express.json());
+
+const data = require('./data.json');
+const notes = data.notes;
+
+app.get('/api/notes', (req, res) => {
+  if (req.params.id === undefined) {
+    res.send(notes);
+  }
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const parsedId = parseInt(req.params.id, 10);
+  if (notes[parsedId] !== undefined) {
+    res.send(notes[parsedId]);
+  } else if (isNaN(parsedId)) {
+    res.status(400).send({
+      error: 'id must be a positive integer'
+    });
+  } else {
+    res.status(400).send({
+      error: `cannot find note with id ${parsedId}`
+    });
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  if (req.body.content !== undefined) {
+    const newNote = {
+      id: data.nextId,
+      content: req.body.content
+    };
+    notes[data.nextId] = newNote;
+    data.nextId++;
+    const dataStr = JSON.stringify(data, null, 2);
+    fs.writeFile('data.json', dataStr, 'utf8', err => {
+      if (err) throw err;
+    });
+    res.status(201).json(newNote);
```

Comment:
We don't know that the note was successfully save until `writeFile` completes, so you will need to move your success responses into your `writeFile` callbacks and only respond with a success if there is no `err`.